easybuild/tools/toolchain/compiler.py

```diff
@@ -278,18 +279,40 @@ def _set_optimal_architecture(self, default_optarch=None):
                                 (--optarch and --optarch=GENERIC still override this value)
         """
         optarch = None
+        use_generic = False
         # --optarch is specified with flags to use
-        if build_option('optarch') is not None and build_option('optarch') != OPTARCH_GENERIC:
-            optarch = build_option('optarch')
-        # --optarch=GENERIC
-        elif build_option('optarch') == OPTARCH_GENERIC:
+        if build_option('optarch') is not None:
+            for optarch_option in build_option('optarch').split(','):
+                # No compilers specified, and option is not generic, so simply take what is there
+                if ":" not in optarch_option and optarch_option != OPTARCH_GENERIC:
+                    optarch = optarch_option
+                    break
+                # Option is generic, so mark it as to be set later and stop iterating
+                elif optarch_option == OPTARCH_GENERIC:
+                    use_generic = True
+                    break
```

Comment:
You are right in the first part of the comment. My mind set was to accept the "first match" and properly document it. Even validating every entry you need something like that to deal with things like `--optarch=Intel:xHost,GCC:something,Intel:xSSE2`.

I think that like it is now is not just more powerful, but also easier to implement (the validation requires a couple of lines more of code, even though it is not significant). It comes with a cost, which is _lack of strictness_ (sounds worse than _flexibility_ :-P). However I have no strong feelings about doing it like you suggest. Just let me know which direction you want to take.

Regarding the parsing bit and the separators: Couldn't we use some similar library to parse that? Even without it, it shouldn't be too complicated to accept something like `--optarch=Intel:'O2 -Wl,-Bstatic',GCC:something`.

Beyond that, do we really need that capability? I mean, none of the compilers I know require it (see list below), and if an user wants to put flags in `--optarch` that don't belong there, should we get out of our way to support that? In my opinion we should support that just if a compiler requires it.

* Intel: `-xHost` or `-xAVX2`, etc
* GCC: `-march=x86-64 -mtune=generic`
* PGI: `-tp=haswell`
* clang: `-march=x86-64 -mtune=generic`
* IBM XL: `-qtune=auto`

And just as I finish typing this I see that `nvcc` accepts things like `--gpu-architecture=compute_50 --gpu-code=sm_50,compute_50` and my argument falls apart and goes down the drain. So, coming back to the previous point, do we accept something like `--optarch=Intel:'O2 -Wl,-Bstatic',GCC:something`, or do we change the characters? Changing them is trivial so I am ok with it.

Last but not least: I strongly oppose the configuration file idea. My arguments are:

1. Disrupts our workflow. We do this sort of tuning with environment variables that are set in configuration modules. We just use files for python code (custom easyblocks, toolchains, MNS)
2. It is kind of overkill. The longest string I can think of is something like `Intel:xCORE-AVX2,GCC:-march=x86-64 -mtune=generic,PGI:-tp=haswell,CLANG:-march=x86-64 -mtune=generic,IBMXL:-qtune=auto,CUDA:--gpu-architecture=compute_50 --gpu-code=sm_50,compute_50`. This is an exaggeration, but even exaggerating it, it is pretty manageable.

I would be willing to accept a file as an *additional* option, if it fits better in people's workflow, but not excluding the `--optarch=Compiler1:flag1,Compiler2:flag2` syntax